tensorflow_probability/python/experimental/mcmc/expectations_reducer.py

```diff
@@ -0,0 +1,227 @@
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+# Dependency imports
+import tensorflow.compat.v2 as tf
+
+from tensorflow_probability.python.experimental.mcmc import reducer as reducer_base
+from tensorflow_probability.python.internal import nest_util
+from tensorflow_probability.python.internal import prefer_static as ps
+from tensorflow_probability.python.mcmc.internal import util as mcmc_util
+from tensorflow.python.util import nest  # pylint: disable=g-direct-tensorflow-import
+
+
+__all__ = [
+    'ExpectationsReducer',
+]
+
+
+ExpectationsReducerState = collections.namedtuple(
+    'ExpectationsReducerState', 'num_samples, running_total')
+
+
+class ExpectationsReducer(reducer_base.Reducer):
+  """`Reducer` that computes a running expectation.
+
+  `ExpectationsReducer` calculates expectation over some arbitrary structure
+  of `callables`. A callable is a function that accepts a Markov chain sample
+  and kernel results, and outputs the relevant value for expectation
+  calculation. In other words, if we denote a callable as f(x,y),
+  `ExpectationsReducer` computes E[f(x,y)] for all provided functions. The
+  finalized expectation will also identically mimic the structure of
+  `callables`.
+
+  As with all reducers, ExpectationsReducer does not hold state information;
+  rather, it stores supplied metadata. Intermediate calculations are held in
+  a state object, which is returned via `initialize` and `one_step` method
+  calls.
+  """
+
+  def __init__(self, callables=lambda sample, kr: sample, name=None):
+    """Instantiates this object.
+
+    Args:
+      callables: A (possibly nested) structure of functions that accept a
+        chain state and kernel results. Defaults to simply returning the
+        incoming chain state.
+      name: Python `str` name prefixed to Ops created by this function.
+        Default value: `None` (i.e., 'expectations_reducer').
+    """
+    self._parameters = dict(
+        callables=callables,
+        name=name or 'expectations_reducer'
+    )
+
+  def initialize(self, initial_chain_state, initial_kernel_results=None):
+    """Initializes an empty `ExpectationsReducerState`.
+
+    Args:
+      initial_chain_state: A (possibly nested) structure of `Tensor`s or Python
+        `list`s of `Tensor`s representing the current state(s) of the Markov
+        chain(s).
+      initial_kernel_results: A (possibly nested) structure of `Tensor`s
+        representing internal calculations made in a related `TransitionKernel`.
+
+    Returns:
+      state: `ExpectationsReducerState` representing a stream of no inputs.
+    """
+    with tf.name_scope(
+        mcmc_util.make_name(self.name, 'expectations_reducer', 'initialize')):
+      initial_chain_state = tf.nest.map_structure(
+          tf.convert_to_tensor,
+          initial_chain_state)
+      return ExpectationsReducerState(
+          num_samples=tf.zeros((), dtype=tf.int32),
```

Comment:
Hmm, why not add a RunningMean to sample_stats.py and use that here? RunningMean is generally useful, and it'd follow the pattern of the CovarianceReducer.

Reply:
Yeah, that's a good point. I thought expectations was trivial enough to not warrant a separate algorithm in sample_stats, but it doesn't hurt!

Reply:
Certainly this algorithm is simple, but if we were to use the same algorithm as running covariance (i.e. `new_mean = state.mean + chunk_n * delta_mean / new_n`), then that's less trivial, and possibly more robust (certainly to overflow).